### PR TITLE
70 invalidate queries

### PR DESCRIPTION
**Pull request by jasoncavanaugh**
## Description of this change
#70 Refactor to use `invalidateQueries` instead of `refetch` when mutations succeed. 
### Before this PR:
\<screenshot(s)>
### After this PR:
\<screenshot(s)>
## Checklist
- [x] Added link to existing Issue (created issue if there wasn't one already)
- [ ] Added screenshots before/after for UI changes
- [x] PR is from branch pushed to this repo so that it will trigger Railway PR deployment
- [x] Code follows the style of this project

### CODE REVIEW

**Comment by jasoncavanaugh** [comment]
So I'm honestly not sure if this is better. It seems to me that the code is a bit less intuitive to a newcomer than it was before (e.g. `refetch` is a lot easier to understand at a glance compared to `invalidateQueries`). Additionally, I had to do some refactoring of the query keys to make it a bit more systematic, since otherwise we would have raw strings lying around that are related to each other. This is probably a good change that we should keep in general. But it also meant that I had to separate out data fetching logic in a few places where I otherwise wouldn't have (in general, I prefer keeping everything inline until it needs to be used somewhere else). Another thing that I'm not sure how I feel about is the fact that, while we no longer need to pass around a `refetch` function as a prop to different components, we still sometimes need to pass around other things instead. For example, if a we had a fetching function that depended on the `eventId`, in order to invalidate that queries cache via the query key, we would need access to the `eventId`, which means that it has to be passed around as a prop. So it's not always the case that we are saving ourselves the trouble of passing around props. So I guess I'm still a bit unsure of this change is actually a win for code readability and maintainability. But it seems that people tend to opt for query invalidation with this library instead of refetching, as indicated by the original article that I linked. Would love to hear any thoughts from @mattsahn @6mp @zuechai

**Comment by mattsahn** [comment]
Thanks, @jasoncavanaugh.  I'm going to try to get some opinion on this one from another frontend/React SME on my team, as well.

**Comment by prndP** [comment]
@jasoncavanaugh  Hi Jason, @mattsahn  asked me to take a look here.

Under perfect circumstances, we should always be computing the "fetch" via a computed key (if we KNOW some API depend on ABC inputs,  then ABC goes into the computation of that key which would automatically be classified as a new fetch). If we do NOT know, then that means either "TIME" or a "USER ACTION" is the trigger, in which case it's perfectly fine to call invalidate.

Looking at this file, (https://github.com/grassrootsgrocery/admin-portal/blob/032e9c98f64158dfc855b0a799113c55ababcc09/client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx) this is clearly falling in the latter use case. The "driver" endpoint, which takes no args, must be invalidated and the invalidation happens due to a user initiated action which makes an "UPDATE" call.  Any writable operation in a set of CRUD endpoints will naturally invalidate reads.

Good. This actually makes total sense from a reasoning perspective (invalidate being much better than refetch, because we are explicitly stating our intention to invalidate BECAUSE of a write). But, for some reason it "feels bad". I think your instincts are good because your intuition is telling you something is not developer friendly with the current implementation. I believe the root cause is actually not with the dependency itself, but where and when invalidation call is happening.

The reason in my opinion of why it feels icky is because here the "data invalidation" concerns are actually being mixed with the "dropdown" concern. Why does setting the onSuccess portion of a dropdown menu invalidate my API call? Whether it's refetch or invalidate, neither makes a lot of sense. But if we see invalidateQuery immediately after another successful API call, the reason becomes much more obvious: the very act of calling the "write" portion of a CRUD api invalidates the "read". This has nothing to do with dropdowns, and 100% to do with synchronizing API states.

Even if you change nothing else, if you move the invalidation piece out of the Dropdown callback up to the useMutation hook, it will feel much more natural. We are declaring "update locations" to be an action which both: writes to the server state, and invalidates the existing read state, and encapsulating it into one function. This also means it becomes far more portable -> at some point the entire useMutation code might be extracted into a generic "updateLocation" custom hook

**Comment by jasoncavanaugh** [comment]
Thank you for the insightful comment, I think that makes a lot of sense. I will look more into the file you linked to and also try to find similar parts of the code where I can apply a similar reasoning.

**Comment by jasoncavanaugh** [comment]
Bumping this for review @6mp @mattsahn @zuechai

**Comment by mattsahn** [comment]
> Bumping this for review @6mp @mattsahn @zuechai

@prndP, please have another look, too.

**Comment by mattsahn** [comment]
> > Bumping this for review @6mp @mattsahn @zuechai
> 
> @prndP, please have another look, too.

Sorry, @jasoncavanaugh, we owe you a review. This one is over my abilities to give a meaningful review so looking for @zuechai  and @6mp to chime in. I know @prndP is on vacation for a couple more weeks.

**Comment by zuechai** [comment]
I can get on this tonight.

File: client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx [hunk 1]
```diff
@@ -69,64 +58,115 @@ const AssignLocationDropdownItem: React.FC<DropdownItemProps> = (
...
-    if (driver.dropoffLocations.some((locId) => locId === locations[i].id)) {
-      output[i] = true;
-    }
-  }
-  return output;
+): DropoffLocationUIState[] {
+  const idsOfDropoffLocationsAssignedToThisDriver = new Set(
+    driver.dropoffLocations
+  );
+  return locations.map((loc) => ({
```
Comment (6mp) [inline client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx:71]: very clean

File: client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx [hunk 2]
```diff
@@ -69,64 +58,115 @@ const AssignLocationDropdownItem: React.FC<DropdownItemProps> = (
...
-      throw new Error(data.message);
-    }
-    return resp.json();
-  };
+      //Set the state for the newly selected/unselected location to loading
+      setDropoffLocationsUIStateList((prev) =>
+        updateDropoffLocationsUIStateList(prev, idx, "isLoading", true)
+      );
+      const url = `/api/volunteers/drivers/assign-location/${driver.id}`;
+      await fetch(url, {
```
Comment (6mp) [inline client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx:123]: The onError in react-query on gets called if an error is thrown in the call. fetch does not throw if the status != 200 like other libs such as axios. 

We should do 
```ts
if (!response.ok) {
  throw Error("Failed in fetch");
}
```

for the onError to work.

Please let me know if I am correct.

File: client/src/pages/ViewEvent/hooks.ts
Comment (6mp) [inline client/src/pages/ViewEvent/hooks.ts:1]: These hooks are very nice, good job!

**Review by 6mp — COMMENTED** [review]
Sorry about the late review; been sick this week. Overall this looks good!

File: client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx (hunk 2 above)
Comment (zuechai) [reply client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx:123]: Good catch. Your suggestion matches the use in this [file](https://github.com/grassrootsgrocery/admin-portal/blob/e54c100d227d2a7615b33a2e4d66784dde10561c/client/src/utils/http.ts).

**Comment by zuechai** [comment]
This is awesome. I can see your concern @jasoncavanaugh about readability, but I think after a new contributor reads the React Query docs it'll feel approachable. I haven't used React Query much, but seeing invalidation in practice like this really displays it's strength and practicality. Sorry, it took long to get this. I needed to read through docs for both React Query and Node Query first to feel confident before giving an opinon. Overall, this is awesome and a huge improvement I think! 

@prndP Your explanation is really helpful. I really like the idea of encapsulating the example you mentioned into its own hook.

File: client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx (hunk 2 above)
Comment (mattsahn) [reply client/src/pages/DriverLocationInfo/AssignLocationDropdown.tsx:123]: @jasoncavanaugh, once this change is in I think we can go ahead and merge. I played around with the deployed PR and everything looks to be working fine. I'll mark as approved for now.

**Review by mattsahn — APPROVED** [review]
Changes LGTM, though should add in the bit of feedback from @6mp too.

**Pull request merged**